### PR TITLE
feat: make thirdpartypasswordless translateable

### PR DESCRIPTION
**Pull request by porcellus**
## Summary of change

Enables providing translations for thirdpartypasswordless

## Related issues

-   #233 

## Test Plan

- Visual testing would be great for this.
- Some tests check copies as well.

## Documentation changes

TODO:

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [x] If added a new recipe interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [x] If I added a new recipe, I also added the recipe entry point into the `size-limit` section of `package.json` with the size limit set to the current size rounded up.

## Remaining TODOs for this PR

-   [ ] Change base branch (or remove tppwless related parts) 
  - I used tppwless as a base branch because I wanted to extract all translations in one go.
-   [ ] Resolve TODOs in code
- [ ] Add test: Send general error with a translation key and make sure that the value of the translation key is rendered (and also FIELD_ERROR in third party).
- [ ] Test how we work together with i18next
- [ ] How will the user override the `t()` function?
- [ ] How does this behave in server side rendering?

### CODE REVIEW

**Comment by rishabhpoddar** [comment]
# Feedback
- [ ] Destination should be 0.18 and not thirdpartyemailpassword
- [ ] The size check if failing - need to change the limits

File: lib/ts/recipe/emailpassword/translations.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { defaultTranslationsCommon } from "../../components/translations";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/translations.ts:1]: Maybe consider moving all the translations into the theme folder of each recipe.

File: examples/for-tests/src/App.js [hunk 2]
```diff
@@ -746,6 +746,14 @@ function getPasswordlessConfigs({ disableDefaultImplementation }) {
             disableDefaultImplementation,
             style: theme.style,
         },
+        translations: {
```
Comment (rishabhpoddar) [inline examples/for-tests/src/App.js:749]: We should consider changing the name to `translationsForLanguages`

Comment (rishabhpoddar) [inline examples/for-tests/src/App.js:749]: We should move `translations` to supertokens.init and not to have a recipe level, and club all the translation related config into one object as opposed to keeping them all in the root level. You can call this `translationsForLanguages` or `languageTranslation`. The fields in the object are:
- `translations` (contains the variable key values for different diff langs
- `t` function override
- `defaultLanguage`
- `translationScope`

Comment (rishabhpoddar) [reply examples/for-tests/src/App.js:749]: If this name is changing, we should also change the name of the functions in the index.ts files

File: lib/ts/index.ts [hunk 3]
```diff
@@ -36,6 +37,14 @@ export default class SuperTokensAPIWrapper {
         return SuperTokens.getRoutingComponent();
     }
 
+    static changeLanguage(language: string): void {
```
Comment (rishabhpoddar) [inline lib/ts/index.ts:40]: Also export useTranslations cause the user might want to use it during override

File: lib/ts/components/translationContext.tsx [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    const translateFunc = useCallback<TranslationFunc>(
+        (key: string, params: Record<string, string> = {}) => {
+            if (userTranslateFunc !== undefined) {
+                return userTranslateFunc(key, params);
+            }
+            const res = translationStore[currentLanguage] && translationStore[currentLanguage][key];
+            const fallback = translationStore[defaultLanguage] && translationStore[defaultLanguage][key];
+
+            if (res === undefined) {
+                console.warn(`Missing translation ${key} from ${currentLanguage}`);
```
Comment (rishabhpoddar) [inline lib/ts/components/translationContext.tsx:43]: Do these only if in dev mode?

File: lib/ts/translationHelpers.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
+// type TranslationContext = Record<string, Record<Record<string, string>>>;
```
Comment (rishabhpoddar) [inline lib/ts/translationHelpers.tsx:1]: Not sure why this comment is here..

File: lib/ts/translationHelpers.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    }
+}
+
+const CURRENT_LANGUAGE_COOKIE_NAME = "sCurrLanguage";
+
+export function saveCurrentLanguage(language: string, cookieDomain: string): void {
+    setFrontendCookie(CURRENT_LANGUAGE_COOKIE_NAME, language, cookieDomain);
+}
+export function getCurrentLanguageFromCookie(): string | null {
+    return getCookieValue(CURRENT_LANGUAGE_COOKIE_NAME);
```
Comment (rishabhpoddar) [inline lib/ts/translationHelpers.tsx:61]: Add a try  catch and if that fails, ignore it and return `null`

File: lib/ts/translationHelpers.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            event,
+            handlerList.filter((h) => h !== handler)
+        );
+    }
+}
+
+const CURRENT_LANGUAGE_COOKIE_NAME = "sCurrLanguage";
+
+export function saveCurrentLanguage(language: string, cookieDomain: string): void {
+    setFrontendCookie(CURRENT_LANGUAGE_COOKIE_NAME, language, cookieDomain);
```
Comment (rishabhpoddar) [inline lib/ts/translationHelpers.tsx:58]: Add a try  catch and if that fails, ignore it and return `null`

File: lib/ts/utils.ts [hunk 8]
```diff
@@ -274,7 +274,115 @@ export class Deferred<T> {
...
+        } else {
+            res[key] = obj2[key];
+        }
+    }
+
+    return res;
+}
+
+// TODO: Reuse from website repo?
+export function normaliseCookieScopeOrThrowError(cookieScope: string): string {
```
Comment (rishabhpoddar) [inline lib/ts/utils.ts:298]: Maybe import from supertokens-website instead.

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/recipe/thirdparty/translations.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        ...defaultTranslationsCommon.en,
+        ...defaultTranslationsEmailVerification.en,
+
+        THIRD_PARTY_SIGN_IN_AND_UP_HEADER_TITLE: "Sign Up / Sign In",
+
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_START: "By continuing, you agree to our",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_TOS: "Terms of Service",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_AND: "and",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_PP: "Privacy Policy",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_END: "",
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/translations.ts:15]: Missing "continue with <provider>" text?

File: lib/ts/recipe/thirdparty/translations.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+export const defaultTranslationsThirdParty = {
+    en: {
+        ...defaultTranslationsCommon.en,
+        ...defaultTranslationsEmailVerification.en,
+
+        THIRD_PARTY_SIGN_IN_AND_UP_HEADER_TITLE: "Sign Up / Sign In",
+
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_START: "By continuing, you agree to our",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_TOS: "Terms of Service",
+        THIRD_PARTY_SIGN_IN_UP_FOOTER_AND: "and",
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/translations.ts:13]: Need to add sapces around the and

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by porcellus** [comment]
This was manually merged into the thirdpartypasswordless PR

**Pull request closed without merging**